Add force detach (#477)
container-storage-interface/spec: spec.md

### Volume Lifecycle

```
   CreateVolume +------------+          DeleteVolume
 +------------->|  CREATED   +------------------------+
 |              |            +<-----------+           |
 |              +---+----^---+            |           |
 |       Controller |    | Controller     |           v
+++         Publish |    | Unpublish      |          +++
|X|          Volume |    | Volume         |          | |
+-+             +---v----+---+            |          +-+
                | NODE_READY |            |
                +---+----^---+            | Node
               Node |    | Node           | Unpublish
            Publish |    | Unpublish      | Volume
             Volume |    | Volume         | (forced)
                +---v----+---+       +----+----------+
                | PUBLISHED  +------>| QUARANTINED_P |
                +------------+       +---------------+

>                    ControllerUnpublishVolume(fenced)

Reviewer: What if node is shutdown in the state of `NODE_READY` ? only in state of `PUBLISHED` , state can be moved to `QUARANTINED_P ` ?

Author: Right, the QUARANTINE_* states are only designed to cover the case where ControllerUnpublishVolume() is used while in a state where it would currently be forbidden, which include PUBLISHED and VOL_READY. If the node is shut down/rebooted, that's a separate case to consider, but without calling ControllerUnpublishVolume(), it should not affect anything.

Reviewer: Thank you for your reply. You mean when the status is not only `PUBLISHED` but also `VOL_READY` , ControllerUnpublishVolume() can be used? If so, in case of bellow figure, arrow to the `QUARANTINED_S` should be come from `VOL_READY` also? I mean, bellow operation will also be allowed?
```
               ControllerUnpublishVolume(fenced)
VOL_READY -----------------------------------------> QUARANTINED_SP 
```

Author: Yeah there should be another arrow in the diagram but it will be very hard to squeeze it into the ASCII art. I will try.

Reviewer 2: @bswartz 

How about using a tool like [this](https://mermaid-js.github.io/mermaid-live-editor/)? 
Just several human readable lines like below will generates an equivalent output to the ASCII art.

```
stateDiagram-v2
    [*] -->  CREATED : CreateVolume
    CREATED --> [*] : DeleteVolume
    CREATED --> NODE_READY : Controller Publish Volume
    NODE_READY --> CREATED : Controller Unpublish Volume
    NODE_READY --> VOL_READY : Node Stage Volume
    VOL_READY --> NODE_READY : Node Unstage Volume
    VOL_READY --> PUBLISHED : Node Publish Volume
    VOL_READY --> QUARANTINED_S : Controller Unpublish Volume (fenced)
    PUBLISHED --> VOL_READY : Node Unpublish Volume
    PUBLISHED --> QUARANTINED_SP : Controller Unpublish Volume (fenced)
    QUARANTINED_SP --> QUARANTINED_S : Node Unpublish Volume(forced)
    QUARANTINED_S --> CREATED : Node Unstage Volume (forced)
```

Author: Completely agree that that's a better way, but changing the format of these diagrams is beyond the scope of my proposal. Maybe submit a separate PR that replaces the ASCII art with more readable figures? Assuming that other PR merges before this, I'd be happy to rebase on top of it and greatly simplify my changes to the state diagram.

Reviewer 2: Agree. I don't mean to block this PR to be merged unless you use the tool. You may only need to consider recreating the diagram with the tool, if you ever need another big change.

Reviewer: @bswartz thank you for your reply. And also, I understand that modifying the diagram is hard.

Unsupported RPCs SHOULD return an appropriate error code that indicates such (e.g. `CALL_NOT_IMPLEMENTED`).
The full list of plugin capabilities is documented in the `ControllerGetCapabilities` and `NodeGetCapabilities` RPCs.

### A Word on Quarantine States

> The purpose of the `QUARANTINE_S`, `QUARANTINE_P`, and `QUARANTINE_SP` states are to enable recovery from node problems.

Reviewer 3: I do not see any other mention/usage of these states?

Author: Look at the ASCII art state transition diagrams.

Reviewer 3: ah - I searched for word and you are using different word in ASCII diagrams. But that explanation is not enough IMO. Do you think that quarantine states should be part of `ControllerUnpublish` responses?

Author: Well all of the states in that ASCII chart are implicit states that the CO is supposed to follow but aren't expressed anywhere in the RPC interface. These new states are meant to be similarly implicit.

Reviewer 4: `QUARANTINE_S`, `QUARANTINE_P`, and `QUARANTINE_SP` are in the diagrams, but I still have difficulty to understand the meeting of those states.

Author: All 3 states refer to situations where a particular node/volume pair is unsafe to use. The difference is which steps need to be performed to return back to the safe state. The _S state requires only unstage to return to normal, the _P state requires only unpublish to return to normal and the _SP state requires both unpublish and unstage to return to normal. These are inversions of the existing states in state diagram, with uncreatively-chosen names (I'm happy to accept better suggestions).

I wanted to emphasize that we're not merely introducing some new flags that yield different behavior, this change relies on a fundamental reordering of operations to arrive at the desired result. When a node is unreachable you have to sort out the problem by talking ONLY to the controller, and the existing CSI spec expressly forbids ControllerUnpublishing until the node side operations are complete.

Reviewer: I agree with @gnufied and @bswartz . explanation of `QUARANTINE_S` , `QUARANTINE_P` , and `QUARANTINE_SP` would be better to added. I suggest adding explanations to `ControllerUnpublishVolume`  section.

Because CSI is designed to be used in distributed systems, it is inevitable that sometimes volumes will become attached to nodes that get stuck or lost, temporarily or permanently.

> Rather than require an administrator to manually clean up such situation, CSI offers a way disconnect a volume from a node "out of order" such that a volume can be disconnected from a problematic node, and safely connected to a different node, and the node can be reliably and safely cleaned up before accessing that volume again, as opposed to the normal path where the node must confirm a volume is disconnected before the controller can unpublish it.  

Reviewer 5: ```suggestion
Rather than require an administrator to manually clean up in such a situation, CSI offers a way to disconnect a volume from a node "out of order" such that a volume can be disconnected from problematic *node A*, and safely connected to a different *node B*, and then *node A* can be reliably and safely cleaned up before accessing that volume again; as opposed to the normal path whereby *node A* must confirm a volume is disconnected before the controller can unpublish it.  
```

Controller Plugin MUST implement this RPC call if it has `PUBLISH_UNPUBLISH_VOLUME` controller capability.
This RPC is a reverse operation of `ControllerPublishVolume`.

> It MUST be called after all `NodeUnstageVolume` and `NodeUnpublishVolume` on the volume are called and succeed unless the plugin has the `UNPUBLISH_FENCE` capability.

Reviewer 6: `UNPUBLISH_FENCE` should be added to `message ControllerServiceCapability`.

Author: Done

Controller Plugin MUST implement this RPC call if it has `PUBLISH_UNPUBLISH_VOLUME` controller capability.
This RPC is a reverse operation of `ControllerPublishVolume`.

> It MUST be called after all `NodeUnstageVolume` and `NodeUnpublishVolume` on the volume are called and succeed unless the plugin has the `UNPUBLISH_FENCE` capability.

Reviewer 5: ```suggestion
It MUST be called after both `NodeUnstageVolume` and `NodeUnpublishVolume` on the volume are called and succeed, unless the plugin has the `UNPUBLISH_FENCE` capability.
```

The Plugin SHOULD perform the work that is necessary for making the volume ready to be consumed by a different node.
The Plugin MUST NOT assume that this RPC will be executed on the node where the volume was previously used.

If the plugin has the `UNPUBLISH_FENCE` capability, the CO MAY specify `fence` as `true`, in which case the SP MUST ensure that the node may no longer access the volume before returning a successful response.
This results in a transition into one of the `QUARANTINE` states where the node must be cleaned up without being able to access the volume like usual.

> This is intended cut off an unreachable node from accessing volumes so those volumes may be safely published to another node.

Reviewer 5: ```suggestion
This is intended to cut off an unreachable node from accessing volumes so those volumes may be safely published to another node.
```

> Once in one of the `QUARANTINE` states the volume MAY NOT be published to that node again until appropriate cleanup has happened using `NodeUnpublishVolume` and `NodeUnstageVolume` (if applicable).

Reviewer 7: This wording makes the calling of `NodeUnpublishVolume` and `NodeUnstageVolume` optional, is that intentional?

In other words, it seems acceptable that the CO calls `ControllerPublish` for a node, only once the node is recovered. So `ControllerPublish` should also do unfencing in case the node was fenced earlier. The CO could wipe its known state of a node (and attached volumes), without trying to call `NodeUnpublishVolume` and `NodeUnstageVolume` after the fencing.

Author: > This wording makes the calling of `NodeUnpublishVolume` and `NodeUnstageVolume` optional, is that intentional?

Well NodeUnstage is always optional because it's capability-controlled. If the capability is present, then you have to use it after NodeUnpublish to get correct results. However, the other alternative is to reboot the whole node and reset its state. All of the concerns we have with the potential for data corruption result from having unflushed write buffers in the kernel. If you reboot, those concerns go away. The point of the forced unpublish/unstage are to provide a path to cleaning up without requiring a reboot.

> In other words, it seems acceptable that the CO calls `ControllerPublish` for a node, only once the node is recovered. So `ControllerPublish` should also do unfencing in case the node was fenced earlier. The CO could wipe its known state of a node (and attached volumes), without trying to call `NodeUnpublishVolume` and `NodeUnstageVolume` after the fencing.

Yes. In fact it's acceptable today for ControllerUnpublish to fence the node in question, it's just not required. Due to that ambiguity, we need this new flag so the CO can specify that it *is* required in certain cases. And because we're allowing a new sequence of operations that was previously illegal (calling ControllerUnpublish before NodeUnpublish) we have to spell out the implications of that.

Reviewer 7: Good, that all makes sense to me.

I'm a little weary that `NodeUnpublishVolume` and `NodeUnstageVolume` are being retried until they succeed (and cause the node to be moved out of quarantine). There seems to be the possibility that these calls never make any progress and a reboot is required. Probably not an issue that needs to be handled here, as it will become the responsibility of the CO to alert the cluster operator that a node is in quarantine and other nodes are taking over parts of the workload.

Author: > I'm a little weary that `NodeUnpublishVolume` and `NodeUnstageVolume` are being retried until they succeed (and cause the node to be moved out of quarantine). There seems to be the possibility that these calls never make any progress and a reboot is required. Probably not an issue that needs to be handled here, as it will become the responsibility of the CO to alert the cluster operator that a node is in quarantine and other nodes are taking over parts of the workload.

I'm not too worried about this. Usually there are more aggressive ways of cleaning up a mount (like the umount -f command). It there aren't, then the driver should not assert the capability, because it could lead to the problem you mention.

My feeling is that, in situations where there aren't ways to force-unmount, it's usually an indication of a missing feature in the underlying protocol or UI. You always want to be able to clean up client state when the server is permanently dead. Any time rebooting is your only option to solve a problem, it's a sign of bad design, and the proper response is to fix the design so that you can force unmount reliably.

> Once in one of the `QUARANTINE` states the volume MAY NOT be published to that node again until appropriate cleanup has happened using `NodeUnpublishVolume` and `NodeUnstageVolume` (if applicable).

Reviewer 3: Thinking from k8s point of view - how will we actually perform the cleanup on the node? Say a node was shutdown with volumes attached to it and hence CO calls `ControllerUnpublishVolume` with `fence: true`. The node comes back:

1. Should node have some taint to prevent scheduling of pods with these volumes?
2. We could implement something during volume reconstruction in kubelet.  But I am still unsure about how will kubelet mark those volumes as "clean".
3. Are we proposing additional state checks in kube-controller-manager that could detect if volume needs to be detached with `fenced: true`? Currently KCM won't detach volumes from unresponsive/shutdown nodes at all.

Reviewer 6: @gnufied Can you help review this KEP?  https://github.com/kubernetes/enhancements/pull/1116

Reviewer 3: Thanks. I did not know that k8s KEP was updated with CSI spec change. I will take a look.

  // This field is OPTIONAL. Refer to the `Secrets Requirements`
  // section on how to use this field.
  map<string, string> secrets = 3 [(csi_secret) = true];
  
  // Indicates SP MUST make the volume inacessible to the node or nodes
  // it is being unpublished from. Any attempt to read or write data
  // to a volume from a node that has been fenced MUST NOT succeed,
  // even if the volume remains staged and/or published on the node.
  // CO MUST NOT set this field to true unless SP has the
  // UNPUBLISH_FENCE controller capability.
  // The SP MAY make the volume inaccessible even when this field is
  // false.
  // This is an OPTIONAL field.

>   bool fence = 4;

Reviewer 8: Add alpha flags for new fields, etc.

Author: Done

  // This field is OPTIONAL. Refer to the `Secrets Requirements`
  // section on how to use this field.
  map<string, string> secrets = 3 [(csi_secret) = true];

>   // Indicates SP MUST make the volume inacessible to the node or nodes

Reviewer 9: @bswartz s/inacessible/inaccessible

>   // Indicates SP MUST make the volume inacessible to the node or nodes

Reviewer 6: inacessible -> inaccessible

>   // it is being unpublished from. Any attempt to read or write data

Reviewer 5: I agree w/ the spirit of this statement. I do wonder how realistic this is, given that CSI doesn't actually govern the data path and it kind of depends on one or both of (a) the underlying OS / driver implementation; (b) firmware running on a node's hardware controller.

Author: Yes the capability is optional for this reason. The hardware needs to be able to distinguish between nodes and deny access to some while allowing access to others, and furthermore needs to be able to revoke access from nodes which are currently have access and are connected. The security models of many storage systems are too primitive or achieve both or even one of these things, or even one of them.

  // to a volume from a node that has been fenced MUST NOT succeed,
  // even if the volume remains staged and/or published on the node.
  // CO MUST NOT set this field to true unless SP has the

>   // UNPUBLISH_FENCE controller capability.

Reviewer 9: One of the main doubt I have here is, what if the SP implement ONLY node capability ? Is that SP driver intended to use the fence capability successfully ?

Author: If the plugin has no controller, then there's nothing for the CO to talk to in the event that a node becomes unreachable. I would expect a CO to use both of these new capabilities together. However, if a plugin can only support UNPUBLISH_FENCE and not the FORCE_UNPUBLISH capability, then it's still possible to use the controller capability to safely move workloads and their attached volumes to a new node in the event of a node failure, and to manage the node cleanup process through a full node reboot. If a plugin supports FORCE_UNPUBLISH only and not UNPUBLISH_FENCE, then it's not particularly useful for anything except more aggressive node cleanup.

Reviewer 9: @bswartz yeah, that clarifies my doubt! imo, its good to document above scenario ( absence of either controller or node capability) in this spec which can avoid some confusion.  On a side note, I was thinking, eventhough its just `FORCE_UNPUBLISH ` capability advertised by the SP driver, its still useful to cut/fence access to volumes . Because in this enhancement  of NODE capability,  its granular to `Volume` Level. Removing access to the volume by black listing certain NODE IPs .etc are possible from SP driver pov if the CO can flag it in the 'next' NODE call for this volume.

Reviewer 10: Thinking a bit more from the pov of a csi driver, which does NOT support controller publish/unpublish, but supports node stage/unstage (e.g CSI driver for GCP filestore), I have a couple of questions:
1. In the current proposal, the driver does not have a way to opt in to leveraging the fence capability (even if there is a way to black list node ips for a given nfs filer), without supporting a controller publish/unpublish. correct?  e.g today there is a controller service PUBLISH_UNPUBLISH capability, but we dont have a controller capability to only support UNPUBLISH with fence (where a regular controller unpublish is a no-op by the driver, but the unpublish with fence=true can quarantine a node-volume pair). no-op controller publish will be costly since it would create extra resources like VA objects in k8s.
2. Second question is around FORCE_UNPUBLISH. I am trying the understand some example scenarios on what condition would the CO trigger an NODE UNPUBLISH/UNSTAGE(with force=true). Does the CO trigger node unstage/unpublish (force=true), only for quarantined volumes?  Because if so, then for a driver which does NOT support controller PUBLISH_UNPUBLISH (thereby no way to quarantine/fence a volume), but supports node stage/publish cannot leverage the force options. (e.g force unmount for nfs, in case of the nfs server unavailability)

Author: >     1. In the current proposal, the driver does not have a way to opt in to leveraging the fence capability (even if there is a way to black list node ips for a given nfs filer), without supporting a controller publish/unpublish. correct?  e.g today there is a controller service PUBLISH_UNPUBLISH capability, but we dont have a controller capability to only support UNPUBLISH with fence (where a regular controller unpublish is a no-op by the driver, but the unpublish with fence=true can quarantine a node-volume pair). no-op controller publish will be costly since it would create extra resources like VA objects in k8s.

Yes, in order for this feature to be useful to COs, there has to be a control path for the CO to tell the SP which nodes may access the volume and which can't. This is the whole purpose of ControllerPublish/Unpublish. In any scheme where the access granting/denying is out of band, you can't automate it from the CO side, and this feature is about automating the recovery from node failures.

>     2. Second question is around FORCE_UNPUBLISH. I am trying the understand some example scenarios on what condition would the CO trigger an NODE UNPUBLISH/UNSTAGE(with force=true). Does the CO trigger node unstage/unpublish (force=true), only for quarantined volumes?  Because if so, then for a driver which does NOT support controller PUBLISH_UNPUBLISH (thereby no way to quarantine/fence a volume), but supports node stage/publish cannot leverage the force options. (e.g force unmount for nfs, in case of the nfs server unavailability)

The CO should use the force flag on unstage/unpublish if it's cleaning up a volume on a node where data has been lost, or will inevitably be lost. Normal behavior on unpublish/unstage is to fail if data loss would occur, and for the CO to keep retrying, so that data loss does not occur by accident. The CO has to make the judgement that data loss is acceptable and inform the SP of that by using the force flag, so that the SP can successfully clean up, discarding any pending data. Situation like this will occur when the CO has chosen to fence the node and move a workload to a different node (in the interest of minimizing workload downtime) or if the storage itself has experienced a data-losing failure and the CO wishes to write off the lost volume by cleaning up the attachment on the node.

      // This enables COs to, for example, fetch per volume
      // condition after a volume is provisioned.
      GET_VOLUME = 12 [(alpha_enum_value) = true];

>       // Indicates the SP supports ControllerUnpublishVolume.fence

Reviewer 5: ```suggestion
      // Indicates the SP supports the ControllerUnpublishVolume.fence
```

This RPC MUST undo the work by the corresponding `NodeStageVolume`.
This RPC SHALL be called by the CO once for each `staging_target_path` that was successfully setup via `NodeStageVolume`.

If the corresponding Controller Plugin has `PUBLISH_UNPUBLISH_VOLUME` controller capability and the Node Plugin has `STAGE_UNSTAGE_VOLUME` capability, the CO MUST guarantee that this RPC is called and returns success before calling `ControllerUnpublishVolume` for the given node and the given volume, unless the Controller Plugin has `UNPUBLISH_FENCE` capability and the Node Plugin has the `FORCE_UNPUBLISH` capability and the `force` flag is `true`.
The CO MUST guarantee that this RPC is called after all `NodeUnpublishVolume` have been called and returned success for the given volume on the given node.

> If the Node Plugin has the `FORCE_UNPUBLISH` capability, the CO MAY specify `force` as `true` in which case the Node Plugin MUST support unstaging volumes even when access has been revoked with `ControllerUnpublishVolume`.

Reviewer 6: `FORCE_UNPUBLISH` should be added to `message NodeServiceCapability`.

Author: Done

This RPC MUST undo the work by the corresponding `NodeStageVolume`.
This RPC SHALL be called by the CO once for each `staging_target_path` that was successfully setup via `NodeStageVolume`.

> If the corresponding Controller Plugin has `PUBLISH_UNPUBLISH_VOLUME` controller capability and the Node Plugin has `STAGE_UNSTAGE_VOLUME` capability, the CO MUST guarantee that this RPC is called and returns success before calling `ControllerUnpublishVolume` for the given node and the given volume, unless the Controller Plugin has `UNPUBLISH_FENCE` capability and the Node Plugin has the `FORCE_UNPUBLISH` capability and the `force` flag is `true`.

Reviewer 5: consistent grammar
```suggestion
If the corresponding Controller Plugin has the `PUBLISH_UNPUBLISH_VOLUME` controller capability and the Node Plugin has the `STAGE_UNSTAGE_VOLUME` capability, the CO MUST guarantee that this RPC is called and returns success before calling `ControllerUnpublishVolume` for the given node and the given volume, unless the Controller Plugin has the `UNPUBLISH_FENCE` capability and the Node Plugin has the `FORCE_UNPUBLISH` capability and the `force` flag is `true`.
```

The CO MUST guarantee that this RPC is called after all `NodeUnpublishVolume` have been called and returned success for the given volume on the given node.

If the Node Plugin has the `FORCE_UNPUBLISH` capability, the CO MAY specify `force` as `true` in which case the Node Plugin MUST support unstaging volumes even when access has been revoked with `ControllerUnpublishVolume`.

> Because data loss is inevitable in such circumstances, the `force` flag is an indication that success is desired even it if means losing data.

Reviewer 6: it if means -> if it means

> Because data loss is inevitable in such circumstances, the `force` flag is an indication that success is desired even if it means losing data.

Reviewer 5: how does a user communicate to a CO that data loss is acceptable? I get the point here, but I'm wondering if we're putting the cart ahead of the horse w/ this feature: if no CO's allow a user to say "yeah, I don't care about data loss here" then ... who's going to be able to use this feature? maybe this was already discussed in k8s SIG storage, and there's a grand plan for a k8s API to deal with this nuance?

> It is essential that after a successful call to `NodeUnstageVolume` that there be no buffered data on the node related to the volume which might result in unintetional modification of the volume if it were to be subsequently re-staged to that node.

Reviewer 6: there be -> there will be
if it were -> if it was

> 

Reviewer 9: s/ unintetional/unintentional

Reviewer 9: @bswartz may be correct this too ?

> It is essential that after a successful call to `NodeUnstageVolume` that there will be no buffered data on the node related to the volume which might result in unintentional modification of the volume if it was to be subsequently re-staged to that node.

Reviewer 5: this talk of "buffered data": again, I agree w/ the spirit of this, but I worry about real life implementation. have we considered what it might look like to write a CSI sanity test for this?

Author: Parts of this are easy to test. We can simulate a situation where the CO (mis)identifies a node as down, and go through the steps of revoking access to a volume from a workload, cleaning up the volume on the node, and reestablishing access to the volume on the same node, and ensuring that the process completes without errors. It would be even better to do this across 2 nodes, so we could simulate the workload moving to another node -- but csi insanity does not lend itself to multi-node testing.

The parts that would be harder to test would be:
* verifying that the workload really lost access to the storage after the fence operations
* verifying that any writes to the volume only came from the new node after a simulated failover
* verifying that the old node didn't corrupt the volume after re-establishing access following a complete and proper cleanup

The basic problem is that CSI itself doesn't give access to low-level-enough information about the volume to reliably determine the absence of these bad cases. We could do some best-effort checking of these things but it's impossible to prove the absence of side effects in a cross-platform way.

This RPC is a reverse operation of `NodePublishVolume`.
This RPC MUST undo the work by the corresponding `NodePublishVolume`.
This RPC SHALL be called by the CO at least once for each `target_path` that was successfully setup via `NodePublishVolume`.

> If the corresponding Controller Plugin has `PUBLISH_UNPUBLISH_VOLUME` controller capability, the CO SHOULD issue all `NodeUnpublishVolume` (as specified above) before calling `ControllerUnpublishVolume` for the given node and the given volume, unless the Controller Plugin has `UNPUBLISH_FENCE` capability and the Node Plugin has the `FORCE_UNPUBLISH` capability and the `force` flag is `true`.

Reviewer 5: ```suggestion
If the corresponding Controller Plugin has the `PUBLISH_UNPUBLISH_VOLUME` controller capability, the CO SHOULD issue `NodeUnpublishVolume` (as specified above) before calling `ControllerUnpublishVolume` for the given node and the given volume, unless the Controller Plugin has the `UNPUBLISH_FENCE` capability and the Node Plugin has the `FORCE_UNPUBLISH` capability and the `force` flag is `true`.
```

The Plugin SHALL assume that this RPC will be executed on the node where the volume is being used.

If the Node Plugin has the `FORCE_UNPUBLISH` capability, the CO MAY specify `force` as `true` in which case the Node Plugin MUST support unpublishing volumes even when access has been revoked with `ControllerUnpublishVolume`.

> Because data loss is inevitable in such circumstances, the `force` flag is an indication that success is desired even it if means losing data.

Reviewer 6: it if means -> if it means

> It is essential that after a successful call to `NodeUnpublishVolume` that there be no buffered data on the node related to the volume which might result in unintetional modification of the volume if it were to be subsequently re-published to that node.

Reviewer 6: there be -> there will be
if it were -> if it was

> It is essential that after a successful call to `NodeUnpublishVolume` that there will be no buffered data on the node related to the volume which might result in unintetional modification of the volume if it was to be subsequently re-published to that node.

Reviewer 5: ```suggestion
It is essential that after a successful call to `NodeUnpublishVolume` that there will be no buffered data on the node related to the volume that might result in unintentional modification of the volume if it was to be subsequently re-published to that node.
```

      // Note that, for alpha, `VolumeCondition` is intended to be
      // informative for humans only, not for automation.
      VOLUME_CONDITION = 4 [(alpha_enum_value) = true];

>       // Indicates that the node supports the NodeUnpublishVolume.force

Reviewer 6: the node supports -> the Node Plugin supports

>       // field. Also indicates that the node supports the

Reviewer 6: the node supports -> the Node Plugin supports
